morpheus/_lib/src/io/loaders/rest.cpp

```diff
@@ -17,21 +17,378 @@
+{
+    net::io_context ioc;
+    tcp::resolver resolver(ioc);
+    beast::tcp_stream stream(ioc);
+    auto const endpoint_results = resolver.resolve(std::string(request[http::field::host]), PORT);
+
+    while (max_retry > 0)
+    {
+        stream.connect(endpoint_results);
+        http::write(stream, request);
+        beast::flat_buffer buffer;
+        http::read(stream, buffer, response);
+        beast::error_code ec;
+        stream.socket().shutdown(tcp::socket::shutdown_both, ec);
+        if (ec && ec != beast::errc::not_connected)
+        {
+            throw beast::system_error{ec};
+        }
+        int status = response.result_int();
+        // Retry only if status is 503 Service Unavailable / 504 Gateway Timeout
+        if (status != 503 && status != 504)
+        {
+            break;
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(retry_interval_milliseconds));
+        retry_interval_milliseconds *= 2;
+        max_retry--;
+    }
+}
+
+py::dict convert_url_query_to_dict(const py::str& raw_query)
+{
+    // Convert query string (param1=true&param2=false) into Python dict ({"param1": "true", "param2": "false"}) for
+    // encoding
+    py::dict params_dict;
+    if (py::len(raw_query) > 0)
+    {
+        py::list split_by_amp = raw_query.attr("split")("&");
+
+        for (auto& query_kv : split_by_amp)
+        {
+            py::list split_by_eq = query_kv.attr("split")("=");
+            if (py::len(split_by_eq) < 2)
+            {
+                throw std::runtime_error("'REST Loader' failed to parse URL query: " + raw_query.cast<std::string>());
+            }
+            params_dict[split_by_eq[0]] = split_by_eq[1];
+        }
+    }
+    return params_dict;
+}
+
+void parse_endpoint_to_url(
+    std::string& endpoint, nlohmann::json& params, std::string& host, std::string& target, std::string& query)
+{
+    py::gil_scoped_acquire gil;
+    py::module_ urllib = py::module::import("urllib.parse");
+
+    // Following the syntax specifications in RFC 1808, urlparse recognizes a netloc only if it is properly
+    // introduced by ‘//’. Otherwise the input is presumed to be a relative URL and thus to start with a path component
+    // ref: https://docs.python.org/3/library/urllib.parse.html
+    // Workaround here is to prepend "http://" if endpoint does not already include one
+    if (endpoint.find("//") == std::string::npos)
+    {
+        endpoint = "http://" + endpoint;
+    }
+
+    py::object result = urllib.attr("urlparse")(endpoint);
+
+    // If not present in URL:
+    //  - host: None
+    //  - path/query: Empty Python String
+    host = result.attr("hostname").is_none() ? "" : result.attr("hostname").cast<std::string>();
+
+    // Encode URL target
+    target = urllib.attr("quote")(result.attr("path")).cast<std::string>();
+    target = target.empty() ? "/" : target;
+
+    py::str raw_query = result.attr("query");
```

Comment:
Could you construct `host`, `target` and `raw_query` only once for multiple params set?